registry/registry_pubkey_update.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's payload")
+	}
+
+	serverSignature, err := hex.DecodeString(data.ServerSignature)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's signature")
+	}
+
+	serverPrivateKey, err := loadServerKeys()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's private key")
+	}
+	serverPubkey := serverPrivateKey.PublicKey
+	serverVerified := verifySignature(serverPayload, serverSignature, &serverPubkey)
+
+	// Overwrite the namespace's public key(s) with the latest keys in the origin
+	if clientVerified && serverVerified {
+		for _, prefix := range data.Prefixes {
+			log.Debug("Start updating namespace: ", prefix)
+
+			// We ensure the prefix exists in registry db via updateNsKeySignChallengeInit function
+
+			existingNs, err := getNamespaceByPrefix(prefix)
+			if err != nil {
+				log.Errorf("Failed to get existing namespace to update: %v", err)
+				return nil, errors.Wrap(err, "Server encountered an error getting existing namespace to update")
+			}
+
+			// Verify the origin is authorized to update
+			registryDbKeySet := jwk.NewSet()
+			err = json.Unmarshal([]byte(existingNs.Pubkey), &registryDbKeySet)
+			if err != nil {
+				log.Errorf("Failed to parse public key as JWKS of registered namespace %s: %v", prefix, err)
+				return nil, errors.Wrapf(err, "Invalid public key format of registered namespace %s", prefix)
+			}
+
+			// Get client's signature from payload
+			keyUpdateAuthzSignature, err := hex.DecodeString(data.KeyUpdateAuthzSignature)
+			if err != nil {
+				return nil, errors.Wrap(err, "Failed to decode the client's key update authorization signature")
+			}
+
+			// Look for the matched key sent back from the client in `registryDbKeySet`
+			matchedKey, found := registryDbKeySet.LookupKeyID(data.MatchedKeyId)
+			if !found {
+				return nil, permissionDeniedError{
+					Message: fmt.Sprintf("The client that tries to update prefix '%s' cannot be authorized because it doesn't contain any public key matching the existing namespace's public key in db", prefix),
+				}
+			}
+
+			rawkey := &ecdsa.PublicKey{}
+			if err := matchedKey.Raw(rawkey); err != nil {
+				return nil, errors.Wrap(err, "failed to generate the raw key from the matched key")
+			}
+
+			keyUpdateAuthzVerified := verifySignature(clientPayload, keyUpdateAuthzSignature, rawkey)
+			if !keyUpdateAuthzVerified {
+				return nil, permissionDeniedError{
+					Message: fmt.Sprintf("The client that tries to update prefix '%s' cannot be authorized because it fails to pass the proof of possession verification", prefix),
+				}
+			}
+
+			// Process origin's latest public key(s)
+			allPubkeyData, err := json.Marshal(data.AllPubkeys)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to convert the latest public key(s) from json to string format for the prefix %s", prefix)
+			}
+			clientPubkeyString := string(allPubkeyData)
```

Comment:
Relying on the string representation seems dicey here.  Do a logical comparison of the public keys, as you do in the `config` module.